### ISSUE TITLE
style attributes are getting duplicated

### ISSUE THREAD

**Issue by gethari** [ueberdosis/tiptap#3908]
### Which packages did you experience the bug in?

core

### What Tiptap version are you using?

2.0.0.rc-1

### What’s the bug you are facing?

We are trying to support `style` tags which can be manipulated by the end users who consume the editor, due to which we are doing following steps to achieve it. But the problem is on the editor's content DOM, the style attributes values are getting duplicated multiple times. We are unsure if this is a bug in the framework, or we are doing something wrong here. 

## Minimal explanation of what we are trying to do

```ts
 addAttributes() {
    return {
      ...this.parent?.(),
      style: {
        default: null,
      },
    };
  },

parseHTML() {
    return [
      {
        tag: "span",
        getAttrs: (dom) => {
          const element = dom;

          return {
            style: element.getAttribute("style"),
          };
        },
      },
    ];
  },
```

Attached an example [Stackblitz](https://stackblitz.com/edit/react-dbfmbq?file=src%2Fstyle.css,src%2FTiptap.js,src%2FMenuBar.js,src%2Findex.js), in which the behavior can be seen by just hitting `Enter` or changing the color of the text inside the Editor.

### What browser are you using?

Chrome

### Code example

https://stackblitz.com/edit/react-dbfmbq?file=src%2Fstyle.css,src%2FTiptap.js,src%2FMenuBar.js,src%2Findex.js

### What did you expect to happen?

Style attributes should not get duplicated.

### Anything to add? (optional)

- We are extending the default's editor extensions such as `Paragraph` / `TextStyle` / & when the color/ background color attribute is getting applied on top of the paragraph, the values are being duplicated
- We could see the same problem happening across all the extensions when we follow the same pattern
- We are unsure if this is a internal bug in tiptap or even maybe from ProseMirror itself,

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [ ] Yes, I’m a sponsor. 💖

**Comment by bdbch** [comment]
Perfect issue, thanks. I'll see that we prioritize it soon!

**Comment by tonyhallett** [comment]
Looks like user error to me.  If it is not then this is why you are getting your result.

The TextStyle, in original form, is a Mark that matches a span with a style attribute ( yours on the other hand matches any span).
```
  parseHTML() {
    return [
      {
        tag: 'span',
        getAttrs: element => {
          const hasStyles = (element as HTMLElement).hasAttribute('style')

          if (!hasStyles) {
            return false
          }

          return {}
        },
      },
    ]
  },
```

Tiptap allows for extensions to add mark ( and node ) attributes, via addGlobalAttributes.  Color is one such extension that adds to the TextStyle mark.
```
export const Color = Extension.create<ColorOptions>({
  name: 'color',

  addOptions() {
    return {
      types: ['textStyle'],
    }
  },

  addGlobalAttributes() {
    return [
      {
        types: this.options.types,
        attributes: {
          color: {

```

TipTap provides an abstraction over the parsing of the dom.

```
schema.parseDOM = parseHTML.map(parseRule => injectExtensionAttributesToParseRule(parseRule, extensionAttributes))
```

Below the getAttrs function provides the attributes for Mark ( or Node ).
You can see that extensions get to provide.  The color extension with parseHTML will provide {color:"red"}
`parseHTML: element => element.style.color?.replace(/['"]+/g, ''),`
You have also added an attribute - this will take the value of the style attribute and will provide 
`{
  style:"color:red"
}`
Finally your parseRule.getAttr also supplies  `{ style: "color:red"}`
```
  parseHTML() {
    return [
      {
        tag: "span",
        getAttrs: (dom) => {
          const element = dom;

          return {
            style: element.getAttribute("style"),
          };
        },
      },
    ];
  },

```
These will be combined to
`{
  color:"red",
  style: "color:red"
}`

```
export function injectExtensionAttributesToParseRule(
  parseRule: ParseRule,
  extensionAttributes: ExtensionAttribute[],
): ParseRule {
  if (parseRule.style) {
    return parseRule
  }

  return {
    ...parseRule,
    getAttrs: node => {
      const oldAttributes = parseRule.getAttrs ? parseRule.getAttrs(node) : parseRule.attrs

      if (oldAttributes === false) {
        return false
      }

      const newAttributes = extensionAttributes.reduce((items, item) => {
        const value = item.attribute.parseHTML
          ? item.attribute.parseHTML(node as HTMLElement)
          : fromString((node as HTMLElement).getAttribute(item.name))

        if (value === null || value === undefined) {
          return items
        }

        return {
          ...items,
          [item.name]: value,
        }
      }, {})

      return { ...oldAttributes, ...newAttributes }
    },
  }
}
```

So given a mark with attributes
`{
  color:"red",
  style: "color:red"
}`
it is then rendered in the editor.

The Tiptap abstraction of toDOM is to provide attributes for you to render.

```
        schema.toDOM = node => renderHTML({
          node,
          HTMLAttributes: getRenderedAttributes(node, extensionAttributes),
        })
```

Your TextStyle renders them 
```
  renderHTML({ HTMLAttributes }) {
    return ["span", mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
  },
```

Again all extension attributes contribute to these dom attributes.

```
export function getRenderedAttributes(
  nodeOrMark: Node | Mark,
  extensionAttributes: ExtensionAttribute[],
): Record<string, any> {
  return extensionAttributes
    .filter(item => item.attribute.rendered)
    .map(item => {
      if (!item.attribute.renderHTML) {
        return {
          [item.name]: nodeOrMark.attrs[item.name],
        }
      }

      return item.attribute.renderHTML(nodeOrMark.attrs) || {}
    })
    .reduce((attributes, attribute) => mergeAttributes(attributes, attribute), {})
}
```

The Color extension uses the color attribute ( which would have come from the parsed dom )
```
            renderHTML: attributes => {
              if (!attributes.color) {
                return {}
              }

              return {
                style: `color: ${attributes.color}`,
              }
            },
```
You TextStyle also supplies as you called addAttributes with a style. This will return the style part from the Mark attrs 
`{
  color:"red",
  style: "color:red"
}`

Thus both Color and your extended TextMark both supply the same attributes.

Attributes are merged, style values are concatenated.  In your instance it will result in {
  style: "color:red;color:red"
}
```
function mergeAttributes(...objects) {
    return objects
        .filter(item => !!item)
        .reduce((items, item) => {
        const mergedAttributes = { ...items };
        Object.entries(item).forEach(([key, value]) => {
            const exists = mergedAttributes[key];
            if (!exists) {
                mergedAttributes[key] = value;
                return;
            }
            if (key === 'class') {
                mergedAttributes[key] = [mergedAttributes[key], value].join(' ');
            }
            else if (key === 'style') {
                mergedAttributes[key] = [mergedAttributes[key], value].join('; ');
            }
            else {
                mergedAttributes[key] = value;
            }
        });
        return mergedAttributes;
    }, {});
}
```

**Comment by nperez0111** [comment]
This has been released with 2.7.0-pre.0 and will be released in tiptap 2.7.0 when this beta has been tested by more users
